src/metadata.rs

```diff
@@ -0,0 +1,145 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+use std::io;
+use xml_writer::XmlWriter;
+
+use cargo::core::{Package, TargetKind};
+use chrono::{DateTime, Utc};
+use serde::Serialize;
+use std::str::FromStr;
+
+use crate::author::Author;
+use crate::component::Component;
+use crate::traits::ToXml;
+
+#[derive(Serialize)]
+pub struct Metadata<'a> {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub timestamp: Option<DateTime<Utc>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub authors: Option<Vec<Author>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub component: Option<Component<'a>>,
+}
+
+impl<'a> Default for Metadata<'a> {
+    fn default() -> Self {
+        Self {
+            timestamp: Some(Utc::now()),
+            authors: None::<Vec<Author>>,
+            component: None::<Component>,
+        }
+    }
+}
+
+impl<'a> From<&'a Package> for Metadata<'a> {
+    fn from(pkg: &'a Package) -> Self {
+        Self {
+            authors: get_authors_from_package(pkg),
+            component: Some(if could_be_application(pkg) {
+                Component::application(pkg).without_scope()
+            } else {
+                Component::library(pkg).without_scope()
+            }),
+            ..Default::default()
+        }
+    }
+}
+
+impl<'a> From<&'a cargo_metadata::Package> for Metadata<'a> {
+    fn from(package: &'a cargo_metadata::Package) -> Self {
+        Self {
+            authors: get_authors_from_package_cm(package),
+            component: Some(if is_an_application(package) {
+                Component::application_cm(package).without_scope()
+            } else {
+                Component::library_cm(package).without_scope()
+            }),
+            ..Default::default()
+        }
+    }
+}
+
+impl ToXml for Metadata<'_> {
+    fn to_xml<W: io::Write>(&self, xml: &mut XmlWriter<W>) -> io::Result<()> {
+        xml.begin_elem("metadata")?;
+
+        if let Some(timestamp) = &self.timestamp {
+            xml.elem_text(
+                "timestamp",
+                &timestamp.to_rfc3339_opts(chrono::SecondsFormat::Secs, true),
+            )?;
+        }
+
+        if let Some(authors) = &self.authors {
+            authors.to_xml(xml)?;
+        }
+
+        if let Some(component) = &self.component {
+            component.to_xml(xml)?;
+        }
+
+        xml.end_elem()
+    }
+}
+
+fn get_authors_from_package<'a>(package: &'a Package) -> Option<Vec<Author>> {
+    let mut authors = vec!();
+
+    for author in package.authors() {
+        authors.push(Author::from_str(author).unwrap());
```

Comment:
It might be better to either have the return be `Result<Option<Vec<Author>>>` or to log the invalid authors and only return the valid ones (something like)

```rust
let mut authors = Vec::new();
let mut invalid_authors = Vec::new();

for author in package.authors() {
    match Author::from_str(author) {
        Ok(author) => authors.push(author),
        Err(e) => invalid_authors.push((author, e)),
    }
}
invalid_authors.into_iter().for_each(|(author, error)| log::debug!("Invalid author {}: {}", author, error))
...
```
(caveat, the above code is artisanal, hand-crafted and may contain trace amounts of compile errors. It should be used as a blueprint only. Also, I'm impatient for [`intersperse`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.intersperse) to stabilize so a dependency on [`itertools`](https://docs.rs/itertools/0.10.1/itertools/fn.join.html) isn't needed to combine this into a single, multi-line debug statement)